File: cache/cache-annotation-processor/src/main/java/ru/tinkoff/kora/cache/annotation/processor/CacheOperation.java
package ru.tinkoff.kora.cache.annotation.processor;

import javax.lang.model.element.ExecutableElement;
import javax.lang.model.element.VariableElement;
import java.util.ArrayList;
import java.util.List;
import java.util.Optional;

public record CacheOperation(Type type, List<String> cacheImplementations, List<VariableElement> parameters, Origin origin) {

    public record Origin(String className, String methodName) {

        @Override
        public String toString() {
            return "[class=" + className + ", method=" + methodName + ']';
        }
    }

    public enum Type {
        GET,
        PUT,
        EVICT,
        EVICT_ALL
    }

    public List<String> getParametersNames(ExecutableElement method) {
        return getParameters(method).stream()
            .map(p -> p.getSimpleName().toString())
            .toList();
    }

    public List<VariableElement> getParameters(ExecutableElement method) {
        if (parameters.isEmpty()) {
            return method.getParameters().stream()
                .filter(this::isParameterSupported)
                .map(p -> ((VariableElement) p))
                .toList();
        } else {
            final List<VariableElement> methodParameters = new ArrayList<>();
            for (var parameter : parameters) {
                final Optional<? extends VariableElement> arg = method.getParameters().stream()
                    .filter(p -> p.getSimpleName().contentEquals(parameter.getSimpleName()))
                    .findFirst();

                if (arg.isPresent()) {
                    methodParameters.add(arg.get());
                } else {
                    throw new IllegalArgumentException("Specified CacheKey parameter '" + parameter + "' is not present in method signature: " + origin());
                }
            }

            return methodParameters;
        }
    }

    public boolean isParameterSupported(VariableElement parameter) {
        return parameters.isEmpty() || parameters.stream().anyMatch(p -> p.getSimpleName().contentEquals(parameter.toString()));
    }
}

File: cache/cache-symbol-processor/src/main/kotlin/ru/tinkoff/kora/cache/symbol/processor/CacheOperation.kt
package ru.tinkoff.kora.cache.symbol.processor

import com.google.devtools.ksp.KspExperimental
import com.google.devtools.ksp.symbol.KSFunctionDeclaration
import com.google.devtools.ksp.symbol.KSValueParameter
import ru.tinkoff.kora.cache.symbol.processor.MethodUtils.Companion.getParameters

@KspExperimental
data class CacheOperation(
    val type: Type,
    val cacheImplementations: List<String>,
    val parameters: List<KSValueParameter>,
    val origin: Origin
) {

    enum class Type {
        GET, PUT, EVICT, EVICT_ALL
    }

    data class Origin(val className: String, val methodName: String) {
        override fun toString(): String = "[class=$className, method=$methodName]"
    }

    fun getParametersNames(method: KSFunctionDeclaration): List<String> {
        return method.getParameters(parameters.asSequence().map { p -> p.name!!.getShortName() }.toList()).asSequence()
            .map { it.name!!.getShortName() }
            .toList()
    }
}
